src/cpu/flexcpu/stld_forwarder.cc

```diff
@@ -0,0 +1,498 @@
+                        inst_ptr->addMemDependency(*other);
+                    }
+                }
+            );
+
+            // NOTE: It is important that this dependency is added after
+            //       the callback above is added to other, since that
+            //       callback needs to resolve first, to make sure that we
+            //       don't accidentally notifyMemReady() before we've added
+            //       all appropriate memory dependencies.
+            DPRINTF(SDCPUDeps, "Dep %d -> %d [mem predicted overlap]\n",
+                    inst_ptr->seqNum(), other->seqNum());
+            inst_ptr->addMemEffAddrDependency(*other);
+        }
+    }
+
+    // END Conservative Memory dependence ordering
+}
+
+void
+StLdForwarder::registerMemBarrier(InflightInst* inst_ptr)
+{
+    // TODO enforce size bound.
+    DPRINTF(SDCPUForwarder, "Inserting Mem Barrier in store buffer (seq %d) "
+                            "Buffer size: %d\n",
+                            inst_ptr->seqNum(),
+                            storeBuffer.size());
+
+    assert(inst_ptr->staticInst()->isMemBarrier());
+    storeBuffer.emplace_back(inst_ptr);
+
+    // As long as register... is always called in-order, instruction squash
+    // happens from the back, and the number of callbacks called is the key.
+    // Same for commit: if in-order, then removing one from front is always
+    // fine.
+    inst_ptr->addSquashCallback([this] {
+        dataMap.erase(storeBuffer.back());
+        storeBuffer.pop_back();
+        assert(!storeBuffer.empty() || dataMap.empty());
+    });
+    inst_ptr->addCommitCallback([this] { commitStore(); });
+}
+
+void
+StLdForwarder::registerStore(InflightInst* inst_ptr)
+{
+    // TODO enforce size bound.
+    DPRINTF(SDCPUForwarder, "Inserting store in store buffer (seq %d) Buffer "
+                            "size: %d\n",
+                            inst_ptr->seqNum(),
+                            storeBuffer.size());
+
+    assert(inst_ptr->staticInst()->isStore());
+    storeBuffer.emplace_back(inst_ptr);
+
+    // As long as registerStore is always called in-order, instruction squash
+    // happens from the back, and the number of callbacks called is the key.
+    // Same for commit: if in-order, then removing one from front is always
+    // fine.
+    inst_ptr->addSquashCallback([this] {
+        dataMap.erase(storeBuffer.back());
+        storeBuffer.pop_back();
+        assert(!storeBuffer.empty() || dataMap.empty());
+    });
+    // TODO consider holding onto the store for a little longer (e.g. until
+    //      memory response time), so more loads can be forwarded. Pretty sure
+    //      the other CPU model LSQs will hold onto items until response anyway
+    //      but we haven't implemented it yet only to keep the code simple for
+    //      now.
+    inst_ptr->addCommitCallback([this] { commitStore(); });
+}
+
+// anonymous namespace for internal linkage only.
```

Comment:
This should be in the header as part of the class description.

Reply:
Why? It's not important to the interface or internal class definition, so I thought it was a good idea to hide it with the function implementation.

Reply:
I moved the declaration into the only function where it's being used, so it's perhaps less weird?